### PR TITLE
Implement NuGetPackageResolver

### PR DESCRIPTION
**Pull request by cston**

### CODE REVIEW

**Comment by cston** [comment]
@tmat, @amcasey, @kevinh-ms, @ManishJayaswal

File: src/Interactive/EditorFeatures/Core/Extensibility/Interactive/InteractiveEvaluator.cs [hunk 1]
```diff
@@ -248,9 +246,12 @@ private Dispatcher Dispatcher
 
         private static MetadataFileReferenceResolver CreateFileResolver(ImmutableArray<string> referencePaths, string baseDirectory)
         {
+            var packagesDirectory = PathUtilities.CombineAbsoluteAndRelativePaths(
+                Environment.GetEnvironmentVariable("userprofile"),
```
Comment (KevinH-MS) [inline src/Interactive/EditorFeatures/Core/Extensibility/Interactive/InteractiveEvaluator.cs]: This can return ```null``` (or empty).  What does nuget do in that case?

File: src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+
+namespace Microsoft.CodeAnalysis.Editor.Interactive
+{
+    internal sealed class NuGetPackageResolverImpl : WORKSPACES::Microsoft.CodeAnalysis.Scripting.NuGetPackageResolver
+    {
+        private readonly string _packagesDirectory;
+
+        internal NuGetPackageResolverImpl(string packagesDirectory)
+        {
+            Debug.Assert(!string.IsNullOrEmpty(packagesDirectory));
```
Comment (KevinH-MS) [inline src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs]: ```IsAbsolute``` checks ```string.IsNullOrEmpty```

File: src/Workspaces/Core/Desktop/Workspaces.Desktop.csproj [hunk 3]
```diff
@@ -186,4 +186,4 @@
     <Import Project="..\..\..\..\build\Targets\VSL.Imports.targets" />
     <Import Project="$(SolutionDir)\.nuget\NuGet.targets" Condition="Exists('$(SolutionDir)\.nuget\NuGet.targets')" />
   </ImportGroup>
-</Project>
+</Project>
```
Comment (KevinH-MS) [inline src/Workspaces/Core/Desktop/Workspaces.Desktop.csproj]: Unnecessary?

File: src/Workspaces/Core/Desktop/packages.config [hunk 4]
```diff
@@ -1,4 +1,4 @@
-<?xml version="1.0" encoding="utf-8"?>
+﻿<?xml version="1.0" encoding="utf-8"?>
 <packages>
   <package id="Microsoft.Composition" version="1.0.27" targetFramework="net45" />
-</packages>
+</packages>
```
Comment (KevinH-MS) [inline src/Workspaces/Core/Desktop/packages.config]: Unnecessary?

File: src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+
+            var tempPath = PathUtilities.CombineAbsoluteAndRelativePaths(Path.GetTempPath(), Guid.NewGuid().ToString("D"));
+            var tempDir = Directory.CreateDirectory(tempPath);
+            try
+            {
+                // Create project.json.
+                var projectJson = PathUtilities.CombineAbsoluteAndRelativePaths(tempPath, "project.json");
+                using (var stream = File.OpenWrite(projectJson))
+                {
+                    using (var writer = new StreamWriter(stream))
```
Comment (Pilchie) [inline src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs]: Convention is to omit the first `{` and indent the multiple `using`s the same level.

File: src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+                var projectLockJson = PathUtilities.CombineAbsoluteAndRelativePaths(tempPath, "project.lock.json");
+                using (var stream = File.OpenRead(projectLockJson))
+                {
+                    using (var reader = new StreamReader(stream))
+                    {
+                        return ReadProjectLockJson(_packagesDirectory, reader);
+                    }
+                }
+            }
+            finally
```
Comment (Pilchie) [inline src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs]: Guessing that you'll need a catch for `IOException` and all the other general IO related stuff here?

File: src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+
+        internal NuGetPackageResolverImpl(string packagesDirectory)
+        {
+            Debug.Assert(!string.IsNullOrEmpty(packagesDirectory));
+            Debug.Assert(PathUtilities.IsAbsolute(packagesDirectory));
+
+            _packagesDirectory = packagesDirectory;
+        }
+
+        internal override ImmutableArray<string> ResolveNuGetPackage(string reference)
```
Comment (Pilchie) [inline src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs]: I don't know how much work this would be, but ideally we could collect all of the `#r`s and `#load's in a single submission/script and resolve them at once, instead of one at a time.

**Comment by Pilchie** [comment]
Seems like a good start to me.

File: src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs (hunk 5 above)
Comment (cston) [inline src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs]: Changed.

File: src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs (hunk 2 above)
Comment (cston) [inline src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs]: Removed extra assert.

File: src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs (hunk 7 above)
Comment (tmat) [inline src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs]: We might be able to do it for #r, but it can't be done for #load since the loaded script may contain #load. We could resolve all #loads within one file.

File: src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+                var projectLockJson = PathUtilities.CombineAbsoluteAndRelativePaths(tempPath, "project.lock.json");
+                using (var stream = File.OpenRead(projectLockJson))
+                using (var reader = new StreamReader(stream))
+                {
+                    return ReadProjectLockJson(_packagesDirectory, reader);
+                }
+            }
+            finally
+            {
+                tempDir.Delete(recursive: true);
```
Comment (KevinH-MS) [inline src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs]: Do we need to try/catch this as well?

File: src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+            using (var jsonWriter = new JsonTextWriter(writer))
+            {
+                jsonWriter.Formatting = Newtonsoft.Json.Formatting.Indented;
+                jsonWriter.WriteStartObject();
+                jsonWriter.WritePropertyName("dependencies");
+                jsonWriter.WriteStartObject();
+                jsonWriter.WritePropertyName(packageName, escape: true);
+                jsonWriter.WriteValue(packageVersion);
+                jsonWriter.WriteEndObject();
+                jsonWriter.WriteEndObject();
```
Comment (KevinH-MS) [inline src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs]: Should we have some logic for determining a reasonable "framework" as well?

Comment (KevinH-MS) [inline src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs]: On second thought, maybe it's not worth it until we plan on running on something other than net40.

File: src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+            }
+            var builder = ArrayBuilder<string>.GetInstance();
+            var targets = (JObject)GetPropertyValue(obj, "targets");
+            // TODO: Choose the appropriate target. For now use the first.
+            foreach (var target in targets)
+            {
+                foreach (var package in (JObject)target.Value)
+                {
+                    var packageRoot = PathUtilities.CombineAbsoluteAndRelativePaths(packagesDirectory, package.Key);
+                    var compile = (JObject)GetPropertyValue((JObject)package.Value, "runtime");
```
Comment (KevinH-MS) [inline src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs]: "runtime"

**Comment by KevinH-MS** [comment]
LGTM (although I'm clueless about the schema of project*json files :smile:)

**Comment by Pilchie** [comment]
Tagging @jasonmalinowski too.

File: src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs (hunk 10 above)
Comment (cston) [inline src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs]: Yes, fixed.

File: src/Interactive/EditorFeatures/Core/Extensibility/Interactive/InteractiveEvaluator.cs (hunk 1 above)
Comment (cston) [inline src/Interactive/EditorFeatures/Core/Extensibility/Interactive/InteractiveEvaluator.cs]: Added check.

File: src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs (hunk 6 above)
Comment (cston) [inline src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs]: Added catches.

File: src/Interactive/EditorFeatures/Core/Extensibility/Interactive/InteractiveEvaluator.cs [hunk 11]
```diff
@@ -248,9 +246,13 @@ private Dispatcher Dispatcher
 
         private static MetadataFileReferenceResolver CreateFileResolver(ImmutableArray<string> referencePaths, string baseDirectory)
         {
+            var userProfilePath = Environment.GetEnvironmentVariable("userprofile");
```
Comment (jasonmalinowski) [inline src/Interactive/EditorFeatures/Core/Extensibility/Interactive/InteractiveEvaluator.cs]: Enviroment.GetFolderPath(Environment.SpecialFolder.UserProfile) instead? I don't think that's guaranteed to exist.

File: src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs [hunk 12]
```diff
@@ -0,0 +1,180 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+extern alias WORKSPACES;
```
Comment (jasonmalinowski) [inline src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs]: Why is this necessary?

File: src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+            {
+            }
+            catch (UnauthorizedAccessException)
+            {
+            }
+            return default(ImmutableArray<string>);
+        }
+
+        /// <summary>
+        /// Syntax is "id/version", matching references in project.lock.json.
```
Comment (jasonmalinowski) [inline src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs]: Not sure if the NuGet team already has a better format for that. The project.lock.json format shouldn't be treated as gospel here.

File: src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+        /// </summary>
+        internal static bool ParsePackageReference(string reference, out string name, out string version)
+        {
+            int separator = reference.IndexOf('/');
+            int n = reference.Length;
+            if ((separator > 0) &&
```
Comment (jasonmalinowski) [inline src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs]: Can we avoid some of the unnecessary parenthesis here? It's making me assume this code is doing more trickery than it is.

File: src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs [hunk 15]
```diff
@@ -0,0 +1,180 @@
...
+                (separator < n - 1) &&
+                (reference.IndexOf('/', separator + 1) < 0))
```
Comment (jasonmalinowski) [inline src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs]: string.Split with max size == 2 might just be easier here. I don't think a single array allocation will kill us here.

File: src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs [hunk 16]
```diff
@@ -0,0 +1,180 @@
...
+                    {
+                        return ReadProjectLockJson(_packagesDirectory, reader);
+                    }
+                }
+                finally
+                {
+                    tempDir.Delete(recursive: true);
+                }
+            }
+            catch (IOException)
```
Comment (jasonmalinowski) [inline src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs]: Can you comment as to why we're tossing exceptions here?

File: src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs (hunk 9 above)
Comment (jasonmalinowski) [inline src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs]: You should just be writing out framework net45 or whatever you're minimum target is. Otherwise NuGet won't know which asset to restore in some cases and will likely choose very wrong. (I'm actually shocked it let you restore at all without one.)

File: src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs [hunk 17]
```diff
@@ -0,0 +1,180 @@
...
+        internal static ImmutableArray<string> ReadProjectLockJson(string packagesDirectory, TextReader reader)
+        {
+            JObject obj;
+            using (var jsonReader = new JsonTextReader(reader))
+            {
+                obj = JObject.Load(jsonReader);
+            }
+            var builder = ArrayBuilder<string>.GetInstance();
+            var targets = (JObject)GetPropertyValue(obj, "targets");
+            // TODO: Choose the appropriate target. For now use the first.
```
Comment (jasonmalinowski) [inline src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs]: Once you specify a framework, then this will be stable. Honestly, hard-coding some arbitrary framework version is probably better than getting lucky like this...

File: src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs [hunk 18]
```diff
@@ -0,0 +1,180 @@
...
+            foreach (var target in targets)
+            {
+                foreach (var package in (JObject)target.Value)
+                {
+                    var packageRoot = PathUtilities.CombineAbsoluteAndRelativePaths(packagesDirectory, package.Key);
+                    var runtime = (JObject)GetPropertyValue((JObject)package.Value, "runtime");
```
Comment (jasonmalinowski) [inline src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs]: Note that:

1. There is no guarantee the runtime group contains managed assemblies as opposed to, say, native assemblies.
2. You'll want the compiler to know about the reference assembly in the case of some packages so you can get XML doc comments. Is there no way to specify both?

File: src/Interactive/EditorFeatures/Core/Extensibility/Interactive/InteractiveEvaluator.cs [hunk 19]
```diff
@@ -248,9 +246,13 @@ private Dispatcher Dispatcher
...
+            var packagesDirectory = (userProfilePath == null) ?
```
Comment (jasonmalinowski) [inline src/Interactive/EditorFeatures/Core/Extensibility/Interactive/InteractiveEvaluator.cs]: There's also a NUGET_PACKAGES environment variable that is respected. You might want to see if there's a flag you can pass to NuGet and restore explicitly rather than hoping your logic is the same as their logic.

File: src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs [hunk 20]
```diff
@@ -0,0 +1,180 @@
...
+                jsonWriter.WriteValue(packageVersion);
+                jsonWriter.WriteEndObject();
+                jsonWriter.WriteEndObject();
+            }
+        }
+
+        /// <summary>
+        /// Read the references from the project.lock.json file.
+        /// </summary>
+        internal static ImmutableArray<string> ReadProjectLockJson(string packagesDirectory, TextReader reader)
```
Comment (jasonmalinowski) [inline src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs]: If this is shipping in our next update, perhaps you can just depend upon my build task? Or perhaps you should one day? I'm not sure.

File: src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs [hunk 21]
```diff
@@ -0,0 +1,180 @@
...
+                PathUtilities.GetDirectoryName(
+                    CorLightup.Desktop.GetAssemblyLocation(typeof(NuGetPackageResolverImpl).GetTypeInfo().Assembly)),
+                "nuget.exe");
+            var startInfo = new ProcessStartInfo()
+            {
+                FileName = nugetExePath,
+                Arguments = $"restore \"{projectJsonPath}\"",
+                CreateNoWindow = true,
+                UseShellExecute = false,
+                RedirectStandardOutput = true,
```
Comment (jasonmalinowski) [inline src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs]: If you redirect StandardOutput and StandardError, you need to have something actively reading both. In your case, the StandardError pipe could block and then you'll deadlock.

File: src/Interactive/EditorFeatures/Core/Extensibility/Interactive/InteractiveEvaluator.cs (hunk 11 above)
Comment (cston) [inline src/Interactive/EditorFeatures/Core/Extensibility/Interactive/InteractiveEvaluator.cs]: Changed, thanks.

File: src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs (hunk 12 above)
Comment (cston) [inline src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs]: The base class (NuGetPackagerResolver) is included in Workspaces.Desktop and Scripting assemblies.

File: src/Interactive/EditorFeatures/Core/Extensibility/Interactive/InteractiveEvaluator.cs (hunk 19 above)
Comment (cston) [inline src/Interactive/EditorFeatures/Core/Extensibility/Interactive/InteractiveEvaluator.cs]: Added explicit "-PackagesDirectory".

File: src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs (hunk 15 above)
Comment (cston) [inline src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs]: Changed to string.Split.

File: src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs (hunk 9 above)
Comment (cston) [inline src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs]: Added "frameworks" : { "net46" : {} }

File: src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs (hunk 21 above)
Comment (cston) [inline src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs]: Added reads for StandardError.

File: src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs (hunk 12 above)
Comment (jasonmalinowski) [inline src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs]: Why is it in both? I wouldn't imagine it lives under Workspaces.Desktop at all.

File: src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs [hunk 22]
```diff
@@ -0,0 +1,204 @@
...
+            var builder = ArrayBuilder<string>.GetInstance();
+            var targets = (JObject)GetPropertyValue(obj, "targets");
+            foreach (var target in targets)
+            {
+                if (target.Key == ProjectLockJsonFramework)
+                {
+                    foreach (var package in (JObject)target.Value)
+                    {
+                        var packageRoot = PathUtilities.CombineAbsoluteAndRelativePaths(packagesDirectory, package.Key);
+                        var runtime = (JObject)GetPropertyValue((JObject)package.Value, "runtime");
```
Comment (jasonmalinowski) [inline src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs]: FYI the runtime section for net46 may not be correct for all cases...we really should have you generating a runtime group as well.

File: src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs [hunk 23]
```diff
@@ -0,0 +1,204 @@
...
+            };
+            _restore(startInfo);
+        }
+
+        private static void NuGetRestore(ProcessStartInfo startInfo)
+        {
+            var process = Process.Start(startInfo);
+            string line;
+            var reader = process.StandardOutput;
+            while ((line = reader.ReadLine()) != null)
```
Comment (jasonmalinowski) [inline src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs]: This still risks deadlock -- if the process starts outputting standard error before standard output, you can still get hit. You need to use asynchrony here.

Comment (cston) [inline src/Interactive/EditorFeatures/Core/Extensibility/Interactive/NuGetPackageResolverImpl.cs]: Added in #5142
